From b15f89849fb387e47c15a271cb05953614b096de Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:39:34 -0500
Subject: [drm] i915: save/restore GMBUS freq across suspend/resume on gen4

Message-id: <1418852380-13061-84-git-send-email-rclark@redhat.com>
Patchwork-id: 102204
O-Subject: [RHEL7 drm 83/89] drm/i915: save/restore GMBUS freq across suspend/resume on gen4
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: drm-next

commit 9f49c37635d5c2a801f7670d5fbf0b25ec461f2c

Author:     Jesse Barnes <jbarnes@virtuousgeek.org>
AuthorDate: Wed Dec 10 12:16:05 2014 -0800
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Mon Dec 15 14:47:33 2014 -0500

    drm/i915: save/restore GMBUS freq across suspend/resume on gen4

    Should probably just init this in the GMbus code all the time, based on
    the cdclk and HPLL like we do on newer platforms.  Ville has code for
    that in a rework branch, but until then we can fix this bug fairly
    easily.

    Bugzilla: https://bugs.freedesktop.org/show_bug.cgi?id=76301
    Signed-off-by: Jesse Barnes <jbarnes@virtuousgeek.org>
    Reviewed-by: Ville Syrjala <ville.syrjala@linux.intel.com>
    Tested-by: Nikolay <mar.kolya@gmail.com>
    Cc: stable@vger.kernel.org
    Signed-off-by: Jani Nikula <jani.nikula@intel.com>

Signed-off-by: Rob Clark <rclark@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 2b80e21..a7eb3eb 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -817,6 +817,7 @@ struct i915_suspend_saved_registers {
  u32 savePIPEB_LINK_N1;
  u32 saveMCHBAR_RENDER_STANDBY;
  u32 savePCH_PORT_HOTPLUG;
+ u16 saveGCDGMBUS;
 };
 
 struct vlv_s0ix_state {
diff --git a/drivers/gpu/drm/i915/i915_reg.h b/drivers/gpu/drm/i915/i915_reg.h
index f548101..a2d35d9 100644
--- a/drivers/gpu/drm/i915/i915_reg.h
+++ b/drivers/gpu/drm/i915/i915_reg.h
@@ -74,6 +74,7 @@
 #define   I915_GC_RENDER_CLOCK_166_MHZ (0 << 0)
 #define   I915_GC_RENDER_CLOCK_200_MHZ (1 << 0)
 #define   I915_GC_RENDER_CLOCK_333_MHZ (4 << 0)
+#define GCDGMBUS 0xcc
 #define PCI_LBPC 0xf4 /* legacy/combination backlight modes, also called LBB */
 
 
diff --git a/drivers/gpu/drm/i915/i915_suspend.c b/drivers/gpu/drm/i915/i915_suspend.c
index 043123c..e22b0e8 100644
--- a/drivers/gpu/drm/i915/i915_suspend.c
+++ b/drivers/gpu/drm/i915/i915_suspend.c
@@ -328,6 +328,10 @@ int i915_save_state(struct drm_device *dev)
   }
  }
 
+ if (IS_GEN4(dev))
+  pci_read_config_word(dev->pdev, GCDGMBUS,
+         &dev_priv->regfile.saveGCDGMBUS);
+
  /* Cache mode state */
  if (INTEL_INFO(dev)->gen < 7)
   dev_priv->regfile.saveCACHE_MODE_0 = I915_READ(CACHE_MODE_0);
@@ -356,6 +360,10 @@ int i915_restore_state(struct drm_device *dev)
  mutex_lock(&dev->struct_mutex);
 
  i915_gem_restore_fences(dev);
+
+ if (IS_GEN4(dev))
+  pci_write_config_word(dev->pdev, GCDGMBUS,
+          dev_priv->regfile.saveGCDGMBUS);
  i915_restore_display(dev);
 
  if (!drm_core_check_feature(dev, DRIVER_MODESET)) {
-- 
1.7.1